cmd/askd/handlers/csv_handler.go

```diff
@@ -0,0 +1,224 @@
+package handlers
+
+import (
+	"bytes"
+	"encoding/csv"
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/coralproject/shelf/internal/ask/form/submission"
+	"gopkg.in/mgo.v2/bson"
+)
+
+// encodeSubmissionsToCSV retrieves the questions and answers on a specific submission.
+// we are returning a [][]string to use the resolut in returing the data in CSV.
+func encodeSubmissionsToCSV(submissions []submission.Submission) ([]byte, error) {
+
+	// Get all the questions for the submissions to the form. It returns a map that contains the widget_id and question.
+	questions := make(map[string]string, 0)
+
+	for _, s := range submissions {
+		for _, r := range s.Answers {
+			questions[r.Question] = r.WidgetID
+		}
+	}
+
+	// Build the header with the columns that we need.
+	header := buildHeader(questions)
+
+	var rows [][]string
+
+	// Add the titles of the header to the CSV.
+	rows = append(rows, getTitles(header))
+
+	// Build the rows.
+	for _, s := range submissions {
+
+		row, err := buildRow(header, s)
+		if err != nil {
+			return nil, err
+		}
+
+		// add the row to the CSV data
+		rows = append(rows, row)
+	}
+
+	// Marshal the data into a CSV string.
+	var buf bytes.Buffer
+	w := csv.NewWriter(&buf)
+
+	w.WriteAll(rows)
+
+	if err := w.Error(); err != nil {
+		return nil, err
+	}
+
+	return buf.Bytes(), nil
+}
+
+// build the header for the CSV.
+// it is returning a slice to get the columns in order when building the CSV
+func buildHeader(questions map[string]string) []map[string]string {
+
+	header := []map[string]string{
+		{
+			"title":    "ID",
+			"widgetID": "",
+		},
+		{
+			"title":    "FormID",
+			"widgetID": "",
+		},
+		{
+			"title":    "Number",
+			"widgetID": "",
+		},
+		{
+			"title":    "Status",
+			"widgetID": "",
+		},
+		{
+			"title":    "Flags",
+			"widgetID": "",
+		},
+		{
+			"title":    "CreatedBy",
+			"widgetID": "",
+		},
+		{
+			"title":    "UpdatedBy",
+			"widgetID": "",
+		},
+		{
+			"title":    "DateCreated",
+			"widgetID": "",
+		},
+		{
+			"title":    "DateUpdated",
+			"widgetID": "",
+		},
+	}
+
+	for k, v := range questions {
+		header = append(header, map[string]string{"title": k, "widgetID": v})
+	}
+
+	return header
+}
+
+// gets the data associated to the header from the submission and build the row
+func buildRow(header []map[string]string, submission submission.Submission) ([]string, error) {
+
+	var row []string
+	var err error
+
+	for i := 0; i < len(header); i++ {
+
+		widgetID := header[i]["widgetID"]
+		title := header[i]["title"]
+
```

Comment:
It is ok because empty values are acceptable here and will be processed correctly in the conditional.